python_modules/dagster/dagster/_core/definitions/assets_job.py

```diff
@@ -213,7 +213,7 @@ def asset2(asset1):
             *(asset_check.node_def for asset_check in asset_checks),
         ]
     else:
-        node_defs = list(filter(None, [asset.node_def for asset in [*source_assets]]))
+        node_defs = list(filter(None, [asset.node_def for asset in [*resolved_source_assets]]))
```

Comment:
I did some exploration here using your code example and now understand what's causing the problem: some of the elements in `source_assets` can have a `node_def` because they're an `AssetsDefinition`, not because they're an observable source asset.

I think the most clear way to filter these out would be to check whether the asset is an observable source asset is `isinstance(asset, SourceAsset) and asset.is_observable`.

Reply:
> I think the most clear way to filter these out would be to check whether the asset is an observable source asset is isinstance(asset, SourceAsset) and asset.is_observable.

Yes, that looks better! pushed the change.